src/lls_model.jl

```diff
@@ -46,50 +178,113 @@ function LLSModel(A :: AbstractMatrix, b :: AbstractVector;
+                  name :: String = "generic-LLSModel"
+                 )
+  nequ = length(b)
+  ncon = length(lcon)
+  A = sparse(Arows, Acols, Avals, nequ, nvar)
+  C = sparse(Crows, Ccols, Cvals, ncon, nvar)
+
+  LLSMatrixModel(A, b, x0=x0, lvar=lvar, uvar=uvar,
+                 C=C, lcon=lcon, ucon=ucon, y0=y0, name=name)
+end
+
+function LLSModel(::Val{:operator},
+                  Arows :: AbstractVector{<: Integer},
                   Acols :: AbstractVector{<: Integer},
                   Avals :: AbstractVector,
                   nvar :: Integer,
                   b :: AbstractVector;
                   x0 :: AbstractVector = zeros(nvar),
-                  lvar :: AbstractVector = fill(-Inf, nvar),
-                  uvar :: AbstractVector = fill(Inf, nvar),
-                  Crows :: AbstractVector{<: Integer} = Int[],
-                  Ccols :: AbstractVector{<: Integer} = Int[],
-                  Cvals :: AbstractVector = Float64[],
-                  lcon :: AbstractVector = Float64[],
-                  ucon :: AbstractVector = Float64[],
-                  y0 :: AbstractVector = zeros(length(lcon)),
+                  lvar :: AbstractVector = fill(-eltype(Avals)(Inf), nvar),
+                  uvar :: AbstractVector = fill(eltype(Avals)(Inf), nvar),
+                  Crows :: AbstractVector{<: Integer} = eltype(Arows)[],
+                  Ccols :: AbstractVector{<: Integer} = eltype(Arows)[],
+                  Cvals :: AbstractVector = eltype(Avals)[],
+                  lcon :: AbstractVector = eltype(Avals)[],
+                  ucon :: AbstractVector = eltype(Avals)[],
+                  y0 :: AbstractVector = zeros(eltype(Avals), length(lcon)),
                   name :: String = "generic-LLSModel"
                  )
 
-  nequ = length(b)
   ncon = length(lcon)
-  if !(ncon == length(ucon) == length(y0))
-    error("The length of lcon, ucon and y0 must be the same")
-  end
+  @lencheck ncon ucon y0
   nnzjF = length(Avals)
-  if !(nnzjF == length(Arows) == length(Acols))
-    error("The length of Arows, Acols and Avals must be the same")
-  end
+  @lencheck nnzjF Arows Acols
   nnzj = length(Cvals)
-  if !(nnzj == length(Crows) == length(Ccols))
-    error("The length of Crows, Ccols and Cvals must be the same")
-  end
+  @lencheck nnzj Crows Ccols
+  nequ = length(b)
 
-  meta = NLPModelMeta(nvar, x0=x0, lvar=lvar, uvar=uvar, ncon=ncon, y0=y0, lin=1:ncon,
-                      nln=Int[], lcon=lcon, ucon=ucon, nnzj=nnzj, nnzh=0, name=name)
+  TA, TC = eltype(Avals), eltype(Acols)
+  Av = zeros(TA, nequ)
+  Atv = zeros(TA, nvar)
+  Aprod(v)  = coo_prod!(Arows, Acols, Avals, v, Av)
+  Atprod(v) = coo_prod!(Acols, Arows, Avals, v, Atv)
+  Cv = zeros(TC, ncon)
+  Ctv = zeros(TC, nvar)
+  Cprod(v)  = coo_prod!(Crows, Ccols, Cvals, v, Cv)
+  Ctprod(v) = coo_prod!(Ccols, Crows, Cvals, v, Ctv)
+  A = LinearOperator{TA}(nequ, nvar, false, false, Aprod, Atprod, Atprod)
```

Comment:
I think this belongs in LinearOperators. What do you think?

Reply:
Then I'm not convinced you really need another constructor here. Just one for abstractmatrix and one for linear ops, maybe?!

Reply:
After COOMatrix is implemented, probably `LinearOperator(A)` will work directly.